torch/csrc/autograd/record_function.cpp

```diff
@@ -191,25 +237,33 @@ inline CallbackManager& manager() {
 } // namespace
 
 bool hasCallbacks() {
-  return manager().hasCallbacks();
+  return manager().hasGlobalCallbacks() || manager().hasThreadLocalCallbacks();
+}
+
+bool hasGlobalCallbacks() {
+  return manager().hasGlobalCallbacks();
+}
+
+bool hasThreadLocalCallbacks() {
+  return manager().hasThreadLocalCallbacks();
 }
 
 void pushCallback(
     std::function<bool(const RecordFunction&)> start,
     std::function<void(const RecordFunction&)> end,
     bool needs_inputs,
     double sampling_prob,
-    std::unordered_set<RecordScope, std::hash<RecordScope>> scopes) {
+    std::unordered_set<RecordScope, std::hash<RecordScope>> scopes,
+    bool is_thread_local) {
```

Comment:
can be minor, but it might be cleaner to just have two separate APIs for thread local callbacks and for global callbacks. Also, your comment below claims that thread local callbacks can't be sampled, so dropping the arg is a good idea

Reply:
agree, will update